File: apps/youtube/extensions/spoof/src/main/java/app/reseam/youtube/spoof/SpoofVideoStreams.java
// SPDX-FileCopyrightText: 2026 AunAli K. <hello@auna.li>
// SPDX-FileCopyrightText: 2026 ReVanced contributors
// SPDX-License-Identifier: GPL-3.0-or-later

package app.reseam.youtube.spoof;

import android.net.Uri;
import android.text.TextUtils;

import java.util.Arrays;
import java.util.Map;

import app.reseam.youtube.core.Logger;
import app.reseam.youtube.core.Settings;

/** Runtime hooks for replacing the app's player response with a compatible client response. */
public final class SpoofVideoStreams {
    private static final String INTERNET_CONNECTION_CHECK_URI_STRING = "https://www.youtube.com/generate_204";
    private static final Uri INTERNET_CONNECTION_CHECK_URI = Uri.parse(INTERNET_CONNECTION_CHECK_URI_STRING);

    private SpoofVideoStreams() {}

    /** VR 1.61 carries AV1 and is used in place of VR 1.43 when AV1 is allowed and AVC is not forced. */
    public static void setClientOrderToUse() {
        ClientType client = ClientType.fromSetting(Settings.getString("spoof_video_streams_client", "android_reel_no_auth"));
        if (client == ClientType.ANDROID_VR_1_43 && Settings.getBoolean("spoof_video_streams_av1", false)
                && !Settings.getBoolean("force_avc_codec", false)) {
            client = ClientType.ANDROID_VR_1_61;
        }
        // Reel requests can take a minute to start playback, so the reel client is used only when chosen.
        StreamingDataRequest.setClientOrder(client,
                Arrays.asList(ClientType.ANDROID_CREATOR, ClientType.ANDROID_VR_1_43, ClientType.VISIONOS),
                true, Settings.getBoolean("force_original_audio", true));
    }

    public static boolean isSpoofingEnabled() {
        return Settings.getBoolean("spoof_video_streams", true);
    }

    public static String rewriteClientContextOsName(String original) {
        if (!isSpoofingEnabled()) return original;
        String rewritten = StreamingDataRequest.getClientOsName();
        if (rewritten != null && !rewritten.equals(original)) {
            Logger.debug(() -> "Client-context OS name spoofed to " + rewritten);
            return rewritten;
        }
        return original;
    }

    public static Uri blockGetWatchRequest(Uri original) {
        if (isSpoofingEnabled() && original != null) {
            String path = original.getPath();
            if (path != null && path.contains("get_watch")) {
                Logger.debug(() -> "Blocking 'get_watch' by returning YouTube connection-check URI");
                return INTERNET_CONNECTION_CHECK_URI;
            }
        }
        return original;
    }

    public static String blockGetAttRequest(String original) {
        if (isSpoofingEnabled() && original != null && containsPath(original, "att/get")) {
            Logger.debug(() -> "Blocking 'att/get' by returning YouTube connection-check URI");
            return INTERNET_CONNECTION_CHECK_URI_STRING;
        }
        return original;
    }

    public static String blockInitPlaybackRequest(String original) {
        if (isSpoofingEnabled() && original != null && containsPath(original, "initplayback")) {
            Logger.debug(() -> "Blocking 'initplayback' by returning YouTube connection-check URI");
            return INTERNET_CONNECTION_CHECK_URI_STRING;
        }
        return original;
    }

    public static void fetchStreams(String url, Map<String, String> requestHeaders) {
        if (!isSpoofingEnabled() || url == null) return;
        try {
            Uri uri = Uri.parse(url);
            String path = uri.getPath();
            if (path == null || !path.contains("player")) return;
            if (path.contains("get_drm_license") || path.contains("heartbeat")
                    || path.contains("refresh") || path.contains("ad_break")) return;
            String videoId = uri.getQueryParameter("id");
            if (videoId == null || videoId.isEmpty()) return;
            StreamingDataRequest.fetchRequest(videoId, requestHeaders);
        } catch (Exception exception) {
            Logger.error(() -> "Spoof stream URL failure: " + exception);
        }
    }

    public static byte[] getStreamingData(String videoId) {
        if (!isSpoofingEnabled() || videoId == null) return null;
        byte[] response = StreamingDataRequest.get(videoId);
        if (response != null) Logger.debug(() -> "Overriding video stream: " + videoId);
        else Logger.debug(() -> "Not overriding streaming data (video stream is null): " + videoId);
        return response;
    }

    public static byte[] removeVideoPlaybackPostBody(Uri uri, int method, byte[] postData) {
        if (isSpoofingEnabled() && method == 2 && uri != null) {
            String path = uri.getPath();
            if (path != null && path.contains("videoplayback")) {
                Logger.debug(() -> "Removed Android video playback POST body for spoofed stream");
                return null;
            }
        }
        return postData;
    }

    public static boolean fixHLSCurrentTime(boolean original) {
        return isSpoofingEnabled() ? false : original;
    }

    public static boolean disableSABR() {
        return isSpoofingEnabled();
    }

    public static boolean useMediaFetchHotConfigReplacement(boolean original) {
        if (original) Logger.debug(() -> "useMediaFetchHotConfigReplacement is set on");
        return isSpoofingEnabled() ? false : original;
    }

    public static boolean usePlaybackStartFeatureFlag(boolean original) {
        if (original) Logger.debug(() -> "usePlaybackStartFeatureFlag is set on");
        return isSpoofingEnabled() ? false : original;
    }

    public static String appendSpoofedClient(String original) {
        if (isSpoofingEnabled() && Settings.getBoolean("spoof_video_streams_stats_for_nerds", true)
                && !TextUtils.isEmpty(original)) {
            return "\u202D" + original + "\u2009(" + StreamingDataRequest.getLastSpoofedClientName() + ")";
        }
        return original;
    }

    private static boolean containsPath(String value, String path) {
        try {
            String originalPath = Uri.parse(value).getPath();
            return originalPath != null && originalPath.contains(path);
        } catch (Exception exception) {
            return false;
        }
    }
}

File: apps/youtube/patch/src/main/kotlin/app/reseam/patches/youtube/misc/SpoofVideoStreams.kt
// SPDX-FileCopyrightText: 2026 AunAli K. <hello@auna.li>
// SPDX-FileCopyrightText: 2026 ReVanced contributors
// SPDX-License-Identifier: GPL-3.0-or-later

package app.reseam.patches.youtube.misc

import app.reseam.patch.ExtClass
import app.reseam.patch.Type
import app.reseam.patch.after
import app.reseam.patch.before
import app.reseam.patch.classTarget
import app.reseam.patch.dex.AccessFlags
import app.reseam.patch.dex.Opcode
import app.reseam.patch.dex.opcode
import app.reseam.patch.dex.parameterTypes
import app.reseam.patch.dex.returnType
import app.reseam.patch.klass
import app.reseam.patch.method
import app.reseam.patch.methods
import app.reseam.patch.patch
import app.reseam.patch.point
import app.reseam.patch.points
import app.reseam.patch.settings.section
import app.reseam.patches.youtube.core.YOUTUBE
import app.reseam.patches.youtube.core.YouTubeSettings
import app.reseam.patches.youtube.core.YouTubeSettingsPages
import app.reseam.patches.youtube.core.mainActivityOnCreate
import app.reseam.patches.youtube.core.youTubeSettings
import app.reseam.patches.youtube.internal.ClientContextEndpoint
import app.reseam.patches.youtube.internal.hookClientContextOsName
import app.reseam.patches.youtube.internal.overrideBooleanFeature
import app.reseam.patches.youtube.internal.playerResponseHook
import app.reseam.patches.youtube.internal.videoIdGetter

private const val STREAMING_DATA =
    "Lcom/google/protos/youtube/api/innertube/StreamingDataOuterClass\$StreamingData;"
private const val BYTE_BUFFER = "java.nio.ByteBuffer"

object SpoofVideoStreams : ExtClass("app.reseam.youtube.spoof.SpoofVideoStreams") {
    val setClientOrderToUse = static("setClientOrderToUse")
    val isSpoofingEnabled = static("isSpoofingEnabled", returns = Type.Boolean)
    val rewriteClientContextOsName = static("rewriteClientContextOsName", Type.String, returns = Type.String)
    val blockGetWatchRequest = static("blockGetWatchRequest", "android.net.Uri", returns = "android.net.Uri")
    val blockGetAttRequest = static("blockGetAttRequest", Type.String, returns = Type.String)
    val blockInitPlaybackRequest = static("blockInitPlaybackRequest", Type.String, returns = Type.String)
    val fetchStreams = static("fetchStreams", Type.String, Type.Map)
    val getStreamingData = static("getStreamingData", Type.String, returns = "[B")
    val removeVideoPlaybackPostBody = static(
        "removeVideoPlaybackPostBody", "android.net.Uri", Type.Int, "[B", returns = "[B",
    )
    val fixHlsCurrentTime = static("fixHLSCurrentTime", Type.Boolean, returns = Type.Boolean)
    val disableSabr = static("disableSABR", returns = Type.Boolean)
    val useMediaFetchHotConfigReplacement = static(
        "useMediaFetchHotConfigReplacement", Type.Boolean, returns = Type.Boolean,
    )
    val usePlaybackStartFeatureFlag = static(
        "usePlaybackStartFeatureFlag", Type.Boolean, returns = Type.Boolean,
    )
    val appendSpoofedClient = static("appendSpoofedClient", Type.String, returns = Type.String)
}

object UserAgentClientSpoof : ExtClass("app.reseam.youtube.spoof.UserAgentClientSpoof") {
    val rewritePackageName = static("rewritePackageName", Type.String, returns = Type.String)
}

object AccountCredentialsInvalidText : ExtClass("app.reseam.youtube.spoof.AccountCredentialsInvalidText") {
    val getOfflineNetworkErrorString = static(
        "getOfflineNetworkErrorString", Type.String, returns = Type.String,
    )
}

private val buildInitPlaybackRequest = method("initplayback request builder") {
    strings("Content-Type", "Range")
    returns("org.chromium.net.UrlRequest\$Builder")
}

private val playerRequestUriBuilder = method("player request URI builder") {
    strings("asig")
    returns("android.net.Uri\$Builder")
}

private val buildRequest = method("Cronet request builder") {
    flags(AccessFlags.PUBLIC or AccessFlags.STATIC)
    calls { owner("Lorg/chromium/net/CronetEngine;"); name("newUrlRequestBuilder") }
    param(1, Type.Map)
    returns("org.chromium.net.UrlRequest\$Builder")
}

private val createStreamingData = method("player response streaming data wrapper") {
    flags(AccessFlags.PUBLIC or AccessFlags.CONSTRUCTOR)
    paramCount(1)
    calls { owner(STREAMING_DATA); name("getDefaultInstance") }
}

private val mediaDataSource = klass("media data source") { strings("DataSpec["); hasInstanceField("[B") }
private val buildMediaDataSource = method("video playback media data source") {
    inClass(mediaDataSource)
    flags(AccessFlags.CONSTRUCTOR)
    hasParam("[B")
}

private val mediaFetchEnumConstructor = method("SABR media-fetch enum constructor") {
    returns(Type.Void)
    strings("DISABLED_BY_SABR_STREAMING_URI")
}

private val nerdsStatsVideoFormatBuilder = method("stats for nerds video format builder") {
    flags(AccessFlags.PUBLIC or AccessFlags.STATIC)
    paramCount(1)
    returns(Type.String)
    strings("codecs=\"")
}

private val protobufParseByteBuffer = method("protobuf byte-buffer parser") {
    name("parseFrom")
    flags(AccessFlags.PROTECTED or AccessFlags.STATIC)
    paramCount(2)
    param(1, BYTE_BUFFER)
    opcode(Opcode.SGET_OBJECT, Opcode.INVOKE_STATIC, Opcode.MOVE_RESULT_OBJECT, Opcode.RETURN_OBJECT)
    custom {
        parameterTypes[0].startsWith("L") && returnType == parameterTypes[0]
    }
}

private val initPlaybackUri = buildInitPlaybackRequest
    .point("initplayback URI toString call") {
        invokeVirtual {
            owner("android.net.Uri")
            name("toString")
            params()
            returns(Type.String)
        }
    }
    .next { resultOf(Type.String) }
    .captureAs("initPlaybackUri", Type.String)

private val playerRequestUris = methods("player request URIs") {
    calls(playerRequestUriBuilder)
}.points {
    invokeVirtual { owner("android.net.Uri\$Builder"); name("build"); params() }
}

private val cronetUrl = buildRequest.point("Cronet request URL") {
    invokeVirtual {
        owner("org.chromium.net.CronetEngine")
        name("newUrlRequestBuilder")
        paramCount(3)
        returns("org.chromium.net.UrlRequest\$Builder")
    }
}.captureArgumentAs("requestUrl", 1, Type.String)

private val streamResponseField = createStreamingData.point("streaming data response field") {
    opcode(Opcode.IPUT_OBJECT)
    field { type(STREAMING_DATA) }
}.field("streaming data response field")

private val streamInputField = createStreamingData.point("player response streaming data field") {
    opcode(Opcode.IGET_OBJECT)
    field { type(STREAMING_DATA) }
}.field("player response streaming data field")

private val playerResponseClass = classTarget("player response protobuf class") {
    bytecode.findClass(streamInputField.owner) ?: error("player response protobuf class is missing")
}

// Follow the app's video-ID getter instead of guessing which protobuf string is the ID.
private val playerResponseModel = klass("player response model") {
    hasInstanceField(playerResponseClass.descriptor)
    implements(videoIdGetter.owner)
}
private val responseVideoId = method("player response video ID") {
    inClass(playerResponseModel)
    name(videoIdGetter.name)
    params()
    returns(Type.String)
}
private val videoDetailsRead = responseVideoId.point("video details read") {
    opcode(Opcode.IGET_OBJECT)
    field { owner(playerResponseClass.descriptor) }
}
private val videoDetailsField = videoDetailsRead.field("video details field")
private val videoIdField = videoDetailsRead.next {
    opcode(Opcode.IGET_OBJECT)
    field { owner(videoDetailsField.type); type(Type.String) }
}.field("video ID field")

private val playerResponseDefaultInstance = playerResponseClass
    .method("<clinit>", inherited = false) {
        params()
        flags(AccessFlags.STATIC)
    }
    .point("player response default instance") {
        opcode(Opcode.SPUT_OBJECT)
        field { type(playerResponseClass.descriptor) }
    }
    .field("player response default instance field")

val spoofVideoStreams = patch("Spoof video streams") {
    description("Requests and installs playback streams from a compatible YouTube client.")
    compatibleWith(YOUTUBE)
    dependsOn(youTubeSettings, playerResponseHook, app.reseam.patches.youtube.internal.clientContextHook)
    settings(
        youTubeSettings,
        section(
            YouTubeSettingsPages.Advanced,
            "Stream compatibility",
            YouTubeSettings.spoofVideoStreams,
            YouTubeSettings.spoofVideoStreamsClient,
            YouTubeSettings.spoofVideoStreamsAv1,
            YouTubeSettings.spoofVideoStreamsStatsForNerds,
        ),
    )

    execute {
        mainActivityOnCreate.before { call(SpoofVideoStreams.setClientOrderToUse) }


        hookClientContextOsName(ClientContextEndpoint.BROWSE, SpoofVideoStreams.rewriteClientContextOsName)
        hookClientContextOsName(ClientContextEndpoint.SEARCH, SpoofVideoStreams.rewriteClientContextOsName)
        hookClientContextOsName(ClientContextEndpoint.REEL, SpoofVideoStreams.rewriteClientContextOsName)

        initPlaybackUri.after {
            capture("initPlaybackUri").assign(
                call(SpoofVideoStreams.blockInitPlaybackRequest, capture("initPlaybackUri")),
            )
        }

        playerRequestUris.forEach {
            next { resultOf("android.net.Uri") }.captureAs("uri", "android.net.Uri").after {
                capture("uri").assign(call(SpoofVideoStreams.blockGetWatchRequest, capture("uri")))
            }
        }

        cronetUrl.before {
            call(SpoofVideoStreams.fetchStreams, capture("requestUrl"), param(1))
            capture("requestUrl").assign(
                call(SpoofVideoStreams.blockGetAttRequest, capture("requestUrl")),
            )
        }

        createStreamingData.after {
            whenFalse(call(SpoofVideoStreams.isSpoofingEnabled)) { returnVoid() }

            val videoDetails = param(0).field(videoDetailsField)
            whenNull(videoDetails) { returnVoid() }
            val videoId = videoDetails.field(videoIdField)
            whenNull(videoId) { returnVoid() }

            val bytes = call(SpoofVideoStreams.getStreamingData, videoId)
            whenNull(bytes) { returnVoid() }

            val parser = call(
                protobufParseByteBuffer,
                staticField(playerResponseDefaultInstance),
                callStatic(BYTE_BUFFER, "wrap", "([B)Ljava/nio/ByteBuffer;", bytes),
            )
            val playerResponse = parser.cast(playerResponseClass.descriptor)
            val replacement = playerResponse.field(streamInputField)
            whenNull(replacement) { returnVoid() }
            thisObject.set(streamResponseField, replacement)
        }

        val mediaUriStore = buildMediaDataSource.point("media request URI store") {
            opcode(Opcode.IPUT_OBJECT)
            field { type("android.net.Uri") }
        }
        val mediaMethodStore = mediaUriStore.next {
            opcode(Opcode.IPUT)
            field { type(Type.Int) }
        }
        val mediaPostDataStore = mediaMethodStore.next {
            opcode(Opcode.IPUT_OBJECT)
            field { type("[B") }
        }
        val mediaUriField = mediaUriStore.field("media request URI field")
        val mediaMethodField = mediaMethodStore.field("media request method field")
        val mediaPostDataField = mediaPostDataStore.field("media request POST body field")
        buildMediaDataSource.after {
            val replacement = call(
                SpoofVideoStreams.removeVideoPlaybackPostBody,
                thisObject.field(mediaUriField),
                thisObject.field(mediaMethodField),
                thisObject.field(mediaPostDataField),
            )
            thisObject.set(mediaPostDataField, replacement)
        }

        nerdsStatsVideoFormatBuilder.after {
            capture("result").assign(call(SpoofVideoStreams.appendSpoofedClient, capture("result")))
        }

        overrideBooleanFeature(45355374L, SpoofVideoStreams.fixHlsCurrentTime)

        val mediaFetchClass = classTarget("SABR media-fetch enum class") {
            bytecode.findClass(mediaFetchEnumConstructor.owner) ?: error("media-fetch enum class is missing")
        }
        val sabrField = mediaFetchEnumConstructor.point("disabled by SABR enum value") {
            string("DISABLED_BY_SABR_STREAMING_URI")
        }.next {
            opcode(Opcode.SPUT_OBJECT)
            field { type(mediaFetchClass.descriptor) }
        }.field("disabled by SABR enum value")
        val sabrMethod = method("SABR enum fallback") {
            returns(mediaFetchClass.descriptor)
            opcodeSequence(Opcode.SGET_OBJECT, Opcode.RETURN_OBJECT)
            custom { parameterTypes.isNotEmpty() }
        }
        sabrMethod.before {
            whenTrue(call(SpoofVideoStreams.disableSabr)) {
                returnValue(staticField(sabrField))
            }
        }

        overrideBooleanFeature(45645570L, SpoofVideoStreams.useMediaFetchHotConfigReplacement)
        overrideBooleanFeature(45665455L, SpoofVideoStreams.usePlaybackStartFeatureFlag)
    }
}
